### ISSUE TITLE
False RedundantCondition when asserting member variable after calling method that modifies it

### ISSUE THREAD

**Issue by stefanfisk** [vimeo/psalm#10148]
Psalm sometimes doesn't understand that method calls might modify instance variables.

This code results in `ERROR: [RedundantCondition](https://psalm.dev/122) - 11:9 - Type 0 for $this->i is always =int(0)`.

```
class Foo {
    private int $i = -1;

    function outer(): void {
        $this->i = 0;

        $this->inner();
        
        assert($this->i === 0);
    }
    
    function inner(): void {
        $this->i++;
    }
}
```

https://psalm.dev/r/75f203c78c

**Comment by boesing** [comment]
Psalm does not have this kind of magic. It does not parse methods to assume something might have happen.
Can you probably show an example where psalm does this kind of magic?

At least I've never seen psalm assuming stuff based on method contents, esp. on non-final classes since inheriting classes could override `inner` and i.e. do `$this->i--`.

AFAIK phpstan does this kind of magic, psalm most likely won't provide this (what is why I prefer psalm over phpstan).

---

By the way, not sure if its just me, but the code makes no sense to me. 
Since `inner` would increment your `0` by `1`, the `assert` will always fail.

**Comment by stefanfisk** [comment]
Or https://psalm.dev/r/ad3bbce7c3

**Comment by boesing** [comment]
> Or https://psalm.dev/r/ad3bbce7c3

That is obviously not working, right? Because it should state that `foo` is not null.

> @boesing it doesn't complain for this quite similar example https://psalm.dev/r/29b4b5695a

That is because using `assert` enforces the value to psalm in the same way `@var` does.
You can also write `/** @var object $this->foo */` which won't make `foo` being an object.

The problem with your example in the initial post is, that it does understand that only `0` is being passed to the property by your assignment.

Same is for your other examples. Psalm does not parse your method, you just narrow the type down by using `assert` which, in your other examples, always is one of the possible values of your provided union type `$foo` property.
It can be either `null` or `string` and by asserting it to be `null` or `not null` you just narrow the union down to a single type.

Psalm gives a damn about what is happening without methods until you provide `/** @psalm-assert <type to be asserted> $this->foo */` to your method.

But this would be almost the same as using `@var` or `assert()` (well, at least `assert` is executed in non-productive systems and will fail while `@var` and `@psalm-assert` won't unless you properly validate your types.

---

TL;DR: in both of your examples, psalm does not care about the methods you are calling. It is **not** aware that the `foo` property contains a literal `foo` string. You can verify that yourself by adding `/** @psalm-trace $this->foo */` after your `inner` method call. If you keep `assert`, it will tell you it is a `string`, if you remove `assert` it will still provide you with `null|string`.

**Comment by weirdan** [comment]
>  I'd just expect psalm to understand that member variables MIGHT be modified by code outside of the current scope

I think you're looking for https://psalm.dev/docs/running_psalm/configuration/#rememberpropertyassignmentsaftercall

**Comment by stefanfisk** [comment]
@weirdan thanks for the tip! it might be. at this point I'm not sure.

**Comment by stefanfisk** [comment]
@boesing Why are you being condescending?

I'll be the first to admit that I do not know much about the inner workings of Psalm. I've skimmed the docs and then learned to use it via trial and error.

When https://psalm.dev/r/261bec6577 does not produce an `RedundantCondition`  while https://psalm.dev/r/735b0c80d8 does it sure looks like a bug to me. Whether or not it is I'll leave up to you all to decide.
